src/librustc_borrowck/borrowck/gather_loans/move_error.rs

```diff
@@ -54,30 +55,40 @@ impl<'tcx> MoveError<'tcx> {
 }
 
 #[derive(Clone)]
-pub struct MoveSpanAndPath {
+pub struct MoveSpanAndPath<'tcx> {
     pub span: syntax_pos::Span,
     pub name: ast::Name,
+    pub pat_source: PatternSource<'tcx>,
 }
 
 pub struct GroupedMoveErrors<'tcx> {
     move_from: mc::cmt<'tcx>,
-    move_to_places: Vec<MoveSpanAndPath>
+    move_to_places: Vec<MoveSpanAndPath<'tcx>>
 }
 
-fn report_move_errors<'a, 'tcx>(bccx: &BorrowckCtxt<'a, 'tcx>,
-                                errors: &Vec<MoveError<'tcx>>) {
+fn report_move_errors<'a, 'tcx>(bccx: &BorrowckCtxt<'a, 'tcx>, errors: &Vec<MoveError<'tcx>>) {
     let grouped_errors = group_errors_with_same_origin(errors);
     for error in &grouped_errors {
         let mut err = report_cannot_move_out_of(bccx, error.move_from.clone());
         let mut is_first_note = true;
-        for move_to in &error.move_to_places {
-            err = note_move_destination(err, move_to.span,
-                                  move_to.name, is_first_note);
-            is_first_note = false;
+	
+	if let Some(pattern_source) = error.move_to_places.get(0){
```

Comment:
Actually, I think I'd just remove the `if let` and merge it with the `match`. So something like this:

```rust
match error.move_to_places.get(0) {
    Some(MoveSpanAndPath { pat_source: PatternSource::LetDecl(_), .. }) => {
        // ignore patterns that are found at the top-level of a `let`;
        // see `get_pattern_source()` for details
    }
    _ => {
        for move_to in &error.move_to_places {
            err = note_move_destination(err, move_to.span, move_to.name, is_first_note);
            is_first_note = false;
        }
    }
}
```

This might read nicer still if we changed the name of the `MoveSpanAndPath` struct into something a bit shorter, like `Place` or `MovePlace`.